### PR TITLE
Feature customer

### PR DESCRIPTION
**Pull request by fvv-vasquez**
Customer classes.
Refactoring methods adding constant.

### CODE REVIEW

File: src/main/java/com/fvv/bookstore/bean/Customer.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * @version 1.0 	
+ *
+ */
+public class Customer extends Person implements Serializable {
+	
+	private static final long serialVersionUID = 1L;
+	private List<String> prodPref;
+
+	/**
+	 * Class constructor specifying the attributes and with super from the Class Person
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/bean/Customer.java:20]: add a space between the comments

File: src/main/java/com/fvv/bookstore/bean/Customer.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	
+	/**
+	 * Class constructor
+	 */
+	public Customer() {
+	}
+
+	/**
+	 * Sets the prodPref.
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/bean/Customer.java]: your variable is called prodPref, however your comment can be something like "Sets the preference products of a customer"

File: src/main/java/com/fvv/bookstore/bean/Customer.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	 * 
+	 * @param prodPref of List type.
+	 */
+	public void setProdPref(List<String> prodPref) {
+		this.prodPref = prodPref;
+	}
+
+	/**
+	 * Gets the last products preferences list.
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/bean/Customer.java]: why it is the last products? i guess you have typos in your description

File: src/main/java/com/fvv/bookstore/bean/Person.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	
+	private Long id;
+	private String name;
+	private String email;
+	private String phone;
+	private String cpf;
+	private Date modificationDate;
+	
+	/**
+	 * Class constructor specifying the attributes
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/bean/Person.java:23]: space between comments

File: src/main/java/com/fvv/bookstore/bean/Person.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	 * Sets the id.
+	 * 
+	 * @param id of Long type.
+	 */
+	public void setId(Long id) {
+		this.id = id;
+	}
+
+	/**
+	 * Gets the last id.
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/bean/Person.java]: why did you add last in most methods?

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * <p>Created on 2018</p>
+ * @version 1.0 	
+ *
+ */
+public final class StringUtil {
+	
+	/**
+	 * Convert a List into a String.
+	 * 
+	 * @param customer of Customer type.
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: wrong parameter

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	 * @return a String of preferences products.
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: please provide details about what the method returns

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	 */
+	public static String convertListToString(final List<String> list) {
+		String preferences = "";
+		for(String pp : list) {
+			preferences += pp + "|";
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: you have a constant for pipe, please use it

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+		}
+		return preferences.substring(0, preferences.length()-1);
+	}
+
+	/**
+	 * Convert a String into a List.
+	 * 
+	 * @param listProducts of String type.
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java:34]: out of date parameters doc, update it

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+	
+	/**
+	 * Convert a List into a String.
+	 * 
+	 * @param customer of Customer type.
+	 * @return a String of preferences products.
+	 */
+	public static String convertListToString(final List<String> list) {
+		String preferences = "";
+		for(String pp : list) {
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: use a more intuitive name for `pp`

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+ */
+public final class StringUtil {
+	
+	/**
+	 * Convert a List into a String.
+	 * 
+	 * @param customer of Customer type.
+	 * @return a String of preferences products.
+	 */
+	public static String convertListToString(final List<String> list) {
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: use a more intuitive name for `list`

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+	/**
+	 * Convert a String into a List.
+	 * 
+	 * @param listProducts of String type.
+	 * @return a List of preferences products.
+	 */
+	public static List<String> convertStringToList(final String listProducts, final String token) {
+		List<String> preferences = new ArrayList<>();
+		String[]products = listProducts.split(token);
+		for(String p : products) {
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: use a more intuitive name for `p`

**Review by cssantos0 — CHANGES_REQUESTED** [review]

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 12 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 11 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 10 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 9 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java:34]: ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 8 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 7 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 6 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: ok

File: src/main/java/com/fvv/bookstore/bean/Customer.java (hunk 1 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/bean/Customer.java:20]: ok

File: src/main/java/com/fvv/bookstore/bean/Customer.java (hunk 2 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/bean/Customer.java]: ok

File: src/main/java/com/fvv/bookstore/bean/Customer.java (hunk 3 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/bean/Customer.java]: ok

File: src/main/java/com/fvv/bookstore/bean/Person.java (hunk 4 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/bean/Person.java:23]: ok

File: src/main/java/com/fvv/bookstore/bean/Person.java (hunk 5 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/bean/Person.java]: ok

File: src/main/java/com/fvv/bookstore/controller/CustomerController.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+	
+	/**
+	 * Search a customer using DAO
+	 * 
+	 * @param id of Long type.
+	 * @return a Customer.
+	 * @throws ControllerException when a problem in controller happens.
+	 * @throws PersonValidationException when a customer is not found.
+	 */
+	Customer findCustomer(final Long id) throws PersonNotFoundException, ControllerException;
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/controller/CustomerController.java]: wrong exception here, fix it , same thing on CellphoneController and laptopController classes in method find. Check there too

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+	/**
+	 * Convert a String into a List.
+	 * 
+	 * @param listProducts of String type.
+	 * @param token of String type
+	 * @return a List of products' preference
+	 */
+	public static List<String> convertStringToList(final String listProducts, final String token) {
+		List<String> preferences = new ArrayList<>();
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: Thinking about this implementation, a better way would be:

```
public static List<String> convertStringToList(final String listProducts, final String token) {
	String[]products = listProducts.split(token);
	return new ArrayList<>(Arrays.asList(products));
}
```

ArrayList has a constructor that accepts a Collection, so you can use Arrays.asList to first convert the array into a list and then passes to the constructor. This is better because avoids a for loop and its iterations.

Please change it

File: src/main/java/com/fvv/bookstore/util/StringUtil.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+	/**
+	 * Convert a List into a String.
+	 * 
+	 * @param listProducts of List type.
+	 * @return a String of products' preference 
+	 */
+	public static String convertListToString(final List<String> listProducts) {
+		String preferences = "";
+		for(String pref : listProducts) {
+			preferences += pref + Constants.PIPE;
```
Comment (cssantos0) [inline src/main/java/com/fvv/bookstore/util/StringUtil.java]: Here it applies the theory about how a String is created we discussed earlier, for this reason, avoid using String concatenation here. Instead, use a StringBuilder and its method append like you already did before

**Review by cssantos0 — CHANGES_REQUESTED** [review]

File: src/main/java/com/fvv/bookstore/controller/CustomerController.java (hunk 13 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/controller/CustomerController.java]: Ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 14 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: Ok

File: src/main/java/com/fvv/bookstore/util/StringUtil.java (hunk 15 above)
Comment (fvv-vasquez) [reply src/main/java/com/fvv/bookstore/util/StringUtil.java]: Ok

**Review by cssantos0 — APPROVED** [review]
